heron/scheduler-core/src/java/com/twitter/heron/scheduler/utils/SchedulerUtils.java

```diff
@@ -39,12 +40,105 @@
 import com.twitter.heron.spi.utils.ShellUtils;
 
 public final class SchedulerUtils {
-  public static final int PORTS_REQUIRED_FOR_EXECUTOR = 9;
+  public static final int PORTS_REQUIRED_FOR_EXECUTOR = ExecutorPorts.getRequiredNumOfPorts();
   public static final int PORTS_REQUIRED_FOR_SCHEDULER = 1;
   public static final String SCHEDULER_COMMAND_LINE_PROPERTIES_OVERRIDE_OPTION = "P";
 
   private static final Logger LOG = Logger.getLogger(SchedulerUtils.class.getName());
 
+  /**
+   * Enum that defines the type of ports that an heron executor needs
+   */
+  public static class ExecutorPorts extends HashMap<ExecutorPorts.ExecutorPortNames, Object> {
+    private static final long serialVersionUID = 3594142561871208879L;
+
+    public enum ExecutorPortNames {
+      MASTER_PORT("master"),
+      TMASTER_CONTROLLER_PORT("tmaster-ctl"),
+      TMASTER_STATS_PORT("tmaster-stats"),
+      SHELL_PORT("shell-port"),
+      METRICS_MANAGER_PORT("metrics-mgr"),
+      SCHEDULER_PORT("scheduler"),
+      METRICS_CACHE_MASTER_PORT("metrics-cache-m"),
+      METRICS_CACHE_STATS_PORT("metrics-cache-s"),
+      CHECKPOINT_MANAGER_PORT("ckptmgr");
+
+      private final String name;
+
+      ExecutorPortNames(String name) {
+        this.name = name;
+      }
+
+      public String toString() {
+        return this.name;
+      }
+
+      public static String[] getPortNames() {
+        List<String> names = new LinkedList<>();
+        for (ExecutorPortNames executorPortNames : ExecutorPortNames.values()) {
+          names.add(executorPortNames.toString());
+        }
+
+        return (String[]) names.toArray();
+      }
+    }
+
+    public static String[] getRequiredPorts() {
+      return new String[]{
+          ExecutorPortNames.MASTER_PORT.toString(),
+          ExecutorPortNames.TMASTER_CONTROLLER_PORT.toString(),
+          ExecutorPortNames.TMASTER_STATS_PORT.toString(),
+          ExecutorPortNames.SHELL_PORT.toString(),
+          ExecutorPortNames.METRICS_MANAGER_PORT.toString(),
+          ExecutorPortNames.SCHEDULER_PORT.toString(),
+          ExecutorPortNames.METRICS_CACHE_MASTER_PORT.toString(),
+          ExecutorPortNames.METRICS_CACHE_STATS_PORT.toString(),
+          ExecutorPortNames.CHECKPOINT_MANAGER_PORT.toString()
+      };
+    }
+
+    public static int getRequiredNumOfPorts() {
+      return getRequiredPorts().length;
+    }
+
+    public static ExecutorPorts withRequiredPorts(List<Integer> portsToUse) {
+      ExecutorPorts executorPorts = new ExecutorPorts();
+      executorPorts.setRequiredPorts(portsToUse);
+      return executorPorts;
+    }
+
+    public void setRequiredPorts(List<Integer> portsToUse) {
+      if (portsToUse.size() < getRequiredNumOfPorts()) {
```

Comment:
Should we check for extra ports provided? i.e. `portsToUse == requiredPorts`

Reply:
We can, but I am not sure it is the place to check for something like that since currently it will work fine even if there are additional ports but not if there are not enough ports.